### PR TITLE
Radix Sort update & Optimization - part 2

### PR DESCRIPTION
**Pull request by KaoCC**
**Note: https://github.com/amdadvtech/Orochi/pull/61 needs to be merged first**

This implementation brings the following improvements and benefits: It is configured with fixed total number of blocks (so it's independent from the input size) + GPU multisplit optimization.  The GPU block size is also configurable now based on the device property.

In summary, this PR does the followings:

- Support setting the the number of thread per block (a.k.a block size) dynamically
- Refactor `exclusiveScanCpu`
- Extend `printKernelInfo`.
 - Support configuring the "NUM_WARPS_PER_BLOCK" as well as the number of warps in a block in the sort kernel
 - Compute the optimal number of inputs for each block to handle.
 - Refactor the usage of stopwatch

### CODE REVIEW

**Comment by KaoCC** [comment]
Old performance records **BEFORE** optimization 

```
executing on AMD Radeon RX 7900 XTX (gfx1100), 48 SIMDs (driverVer.:50500000)
 2.98ms (0.05GKeys/s) sorting 0.2Mkeys [keyValue]
 1.32ms (0.12GKeys/s) sorting 0.2Mkeys [keyValue]
 1.00ms (0.16GKeys/s) sorting 0.2Mkeys [keyValue]
 0.98ms (0.16GKeys/s) sorting 0.2Mkeys [keyValue]
passed: 160.00K keys
 4.94ms (0.32GKeys/s) sorting 1.6Mkeys [keyValue]
 3.66ms (0.44GKeys/s) sorting 1.6Mkeys [keyValue]
 2.96ms (0.54GKeys/s) sorting 1.6Mkeys [keyValue]
 2.66ms (0.60GKeys/s) sorting 1.6Mkeys [keyValue]
passed: 1600.00K keys
15.42ms (1.04GKeys/s) sorting 16.0Mkeys [keyValue]
13.70ms (1.17GKeys/s) sorting 16.0Mkeys [keyValue]
13.75ms (1.16GKeys/s) sorting 16.0Mkeys [keyValue]
13.69ms (1.17GKeys/s) sorting 16.0Mkeys [keyValue]
passed: 16000.00K keys
 2.58ms (0.06GKeys/s) sorting 0.2Mkeys [key]
 2.32ms (0.07GKeys/s) sorting 0.2Mkeys [key]
 2.12ms (0.08GKeys/s) sorting 0.2Mkeys [key]
 1.66ms (0.10GKeys/s) sorting 0.2Mkeys [key]
passed: 160.00K keys
 4.48ms (0.36GKeys/s) sorting 1.6Mkeys [key]
 3.57ms (0.45GKeys/s) sorting 1.6Mkeys [key]
 3.18ms (0.50GKeys/s) sorting 1.6Mkeys [key]
 2.65ms (0.60GKeys/s) sorting 1.6Mkeys [key]
passed: 1600.00K keys
15.06ms (1.06GKeys/s) sorting 16.0Mkeys [key]
11.64ms (1.37GKeys/s) sorting 16.0Mkeys [key]
11.31ms (1.41GKeys/s) sorting 16.0Mkeys [key]
11.30ms (1.42GKeys/s) sorting 16.0Mkeys [key]
passed: 16000.00K keys
>> testing 16 bit sort
 1.79ms (0.09GKeys/s) sorting 0.2Mkeys [keyValue]
 1.25ms (0.13GKeys/s) sorting 0.2Mkeys [keyValue]
 1.22ms (0.13GKeys/s) sorting 0.2Mkeys [keyValue]
 1.08ms (0.15GKeys/s) sorting 0.2Mkeys [keyValue]
passed: 160.00K keys
 2.80ms (0.57GKeys/s) sorting 1.6Mkeys [keyValue]
 2.23ms (0.72GKeys/s) sorting 1.6Mkeys [keyValue]
 1.20ms (1.33GKeys/s) sorting 1.6Mkeys [keyValue]
 1.15ms (1.40GKeys/s) sorting 1.6Mkeys [keyValue]
passed: 1600.00K keys
 9.29ms (1.72GKeys/s) sorting 16.0Mkeys [keyValue]
 7.78ms (2.06GKeys/s) sorting 16.0Mkeys [keyValue]
 7.76ms (2.06GKeys/s) sorting 16.0Mkeys [keyValue]
 7.80ms (2.05GKeys/s) sorting 16.0Mkeys [keyValue]
passed: 16000.00K keys
 1.58ms (0.10GKeys/s) sorting 0.2Mkeys [key]
 1.24ms (0.13GKeys/s) sorting 0.2Mkeys [key]
 1.16ms (0.14GKeys/s) sorting 0.2Mkeys [key]
 1.15ms (0.14GKeys/s) sorting 0.2Mkeys [key]
passed: 160.00K keys
 2.32ms (0.69GKeys/s) sorting 1.6Mkeys [key]
 2.18ms (0.73GKeys/s) sorting 1.6Mkeys [key]
 1.91ms (0.84GKeys/s) sorting 1.6Mkeys [key]
 1.08ms (1.49GKeys/s) sorting 1.6Mkeys [key]
passed: 1600.00K keys
 8.72ms (1.83GKeys/s) sorting 16.0Mkeys [key]
 6.03ms (2.65GKeys/s) sorting 16.0Mkeys [key]
 5.79ms (2.76GKeys/s) sorting 16.0Mkeys [key]
 5.81ms (2.76GKeys/s) sorting 16.0Mkeys [key]
passed: 16000.00K keys
>> done


```

**Comment by KaoCC** [comment]
New performance records **AFTER** optimization

```
executing on AMD Radeon RX 7900 XTX (gfx1100), 48 SIMDs (driverVer.:50500000)
 0.27ms (0.59GKeys/s) sorting 0.2Mkeys [keyValue]
 0.22ms (0.71GKeys/s) sorting 0.2Mkeys [keyValue]
 0.23ms (0.69GKeys/s) sorting 0.2Mkeys [keyValue]
 0.22ms (0.73GKeys/s) sorting 0.2Mkeys [keyValue]
passed: 160.00K keys
 0.56ms (2.86GKeys/s) sorting 1.6Mkeys [keyValue]
 0.52ms (3.05GKeys/s) sorting 1.6Mkeys [keyValue]
 0.55ms (2.94GKeys/s) sorting 1.6Mkeys [keyValue]
 0.55ms (2.92GKeys/s) sorting 1.6Mkeys [keyValue]
passed: 1600.00K keys
 5.98ms (2.67GKeys/s) sorting 16.0Mkeys [keyValue]
 5.97ms (2.68GKeys/s) sorting 16.0Mkeys [keyValue]
 5.97ms (2.68GKeys/s) sorting 16.0Mkeys [keyValue]
 6.01ms (2.66GKeys/s) sorting 16.0Mkeys [keyValue]
passed: 16000.00K keys
 0.27ms (0.58GKeys/s) sorting 0.2Mkeys [key]
 0.19ms (0.85GKeys/s) sorting 0.2Mkeys [key]
 0.18ms (0.88GKeys/s) sorting 0.2Mkeys [key]
 0.23ms (0.71GKeys/s) sorting 0.2Mkeys [key]
passed: 160.00K keys
 0.50ms (3.23GKeys/s) sorting 1.6Mkeys [key]
 0.46ms (3.49GKeys/s) sorting 1.6Mkeys [key]
 0.45ms (3.54GKeys/s) sorting 1.6Mkeys [key]
 0.47ms (3.42GKeys/s) sorting 1.6Mkeys [key]
passed: 1600.00K keys
 3.63ms (4.41GKeys/s) sorting 16.0Mkeys [key]
 3.59ms (4.45GKeys/s) sorting 16.0Mkeys [key]
 3.59ms (4.45GKeys/s) sorting 16.0Mkeys [key]
 3.57ms (4.48GKeys/s) sorting 16.0Mkeys [key]
passed: 16000.00K keys
>> testing 16 bit sort
 0.23ms (0.70GKeys/s) sorting 0.2Mkeys [keyValue]
 0.16ms (1.00GKeys/s) sorting 0.2Mkeys [keyValue]
 0.16ms (1.00GKeys/s) sorting 0.2Mkeys [keyValue]
 0.14ms (1.17GKeys/s) sorting 0.2Mkeys [keyValue]
passed: 160.00K keys
 0.37ms (4.30GKeys/s) sorting 1.6Mkeys [keyValue]
 0.34ms (4.64GKeys/s) sorting 1.6Mkeys [keyValue]
 0.37ms (4.29GKeys/s) sorting 1.6Mkeys [keyValue]
 0.36ms (4.45GKeys/s) sorting 1.6Mkeys [keyValue]
passed: 1600.00K keys
 4.46ms (3.59GKeys/s) sorting 16.0Mkeys [keyValue]
 4.45ms (3.59GKeys/s) sorting 16.0Mkeys [keyValue]
 4.45ms (3.59GKeys/s) sorting 16.0Mkeys [keyValue]
 4.48ms (3.57GKeys/s) sorting 16.0Mkeys [keyValue]
passed: 16000.00K keys
 0.23ms (0.68GKeys/s) sorting 0.2Mkeys [key]
 0.14ms (1.11GKeys/s) sorting 0.2Mkeys [key]
 0.14ms (1.14GKeys/s) sorting 0.2Mkeys [key]
 0.12ms (1.33GKeys/s) sorting 0.2Mkeys [key]
passed: 160.00K keys
 0.27ms (5.98GKeys/s) sorting 1.6Mkeys [key]
 0.25ms (6.32GKeys/s) sorting 1.6Mkeys [key]
 0.24ms (6.54GKeys/s) sorting 1.6Mkeys [key]
 0.25ms (6.45GKeys/s) sorting 1.6Mkeys [key]
passed: 1600.00K keys
 2.01ms (7.94GKeys/s) sorting 16.0Mkeys [key]
 1.97ms (8.12GKeys/s) sorting 16.0Mkeys [key]
 1.98ms (8.08GKeys/s) sorting 16.0Mkeys [key]
 1.99ms (8.04GKeys/s) sorting 16.0Mkeys [key]
passed: 16000.00K keys
>> done

```

File: ParallelPrimitives/RadixSortKernels.h [hunk 1]
```diff
@@ -762,3 +614,131 @@ extern "C" __global__ void ParallelExclusiveScanAllWG( int* gCount, int* gHistog
...
+	{
+		const u32 itemIndex = blockIdx.x * gNItemsPerWG + i;
+
+		const auto item = gSrcKey[itemIndex];
+		const u32 bucketIndex = getMaskedBits( item, START_BIT );
+
+		const int warp = threadIdx.x / 32;
+		const int lane = threadIdx.x % 32;
+
+		__syncthreads();
```
Comment (KaoCC) [inline ParallelPrimitives/RadixSortKernels.h:677]: Note: `syncwarp()` should suffice here but it's not available on HIP.

Comment (takahiroharada) [reply ParallelPrimitives/RadixSortKernels.h:677]: I think someone else suggested but how about making a macro switching these depending on AMD or NV?

**Review by takahiroharada — APPROVED** [review]

**Comment by takahiroharada** [comment]
Looks good. Thank you @KaoCC  and @AtsushiYoshimura0302 .

**Pull request merged**